Register/Add replicas based on UUIDs instead of IP Addresses

kmjayadeep opened the pull request:
Register/Add replicas based on UUIDs.

Please give feedback so that I can proceed further

TODO:

- Save UUIDs on file so that it will be unique across restarts - **DONE**
- Write more Integration tests

Signed-off-by: Jayadeep KM <kmjayadeep@gmail.com>

utkarshmani1997 commented on ci/start_init_test.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-set -x
+# set -x
```
uncomment it.

kmjayadeep replied:
I will, after travis build passes successfully. Now it is difficult to understand the logs as it is too verbose

utkarshmani1997 commented on controller/client/controller_client.go:
```diff
@@ -191,9 +191,11 @@ func (c *ControllerClient) GetVolume() (*rest.Volume, error) {
 
-func (c *ControllerClient) Register(address string, revisionCount int64, replicaType string, upTime time.Duration, state string) error {
+func (c *ControllerClient) Register(address string, uuid string, revisionCount int64, replicaType string, upTime time.Duration, state string) error {
+	fmt.Printf("Registering %v %v", address, uuid)
```
replace fmt.Printf with logrus.Infof

utkarshmani1997 commented on replica/replica.go:
```diff
@@ -59,6 +59,9 @@ type Replica struct {
 	peerCache types.PeerDetails
 	peerFile  *sparse.DirectFileIoProcessor
 
+	uuidLock sync.Mutex
```
Is it being accessed concurrently ? If yes, can you mention about the same in the comment. I am guessing it's single threaded.

kmjayadeep replied:
No, I saw a lock in revision_counter and thought it's required here also. I will remove it

utkarshmani1997 commented on replica/replica.go:
```diff
@@ -243,6 +248,32 @@ func IsHeadDisk(diskName string) bool {
 	return false
 }
 
+func (r *Replica) setUUID() string {
```
can we rename it as readUUID() ?

kmjayadeep replied:
setUUID is creating a new uuid and writing to disk. If it is already there, it'll be read and r.UUID will be set.
We don't have a dedicated method to read UUID

utkarshmani1997 commented on replica/replica.go:
```diff
@@ -243,6 +248,32 @@ func IsHeadDisk(diskName string) bool {
+	if r.UUID == "" {
+		const uuidFileName = "uuid"
+		buf := make([]byte, 36)
+		if _, err := os.Stat(r.diskPath(uuidFileName)); err != nil {
```
can we be more specific about the error ? 
for exp: os.IsNotExist(err)

utkarshmani1997 commented on replica/replica.go:
```diff
@@ -243,6 +248,32 @@ func IsHeadDisk(diskName string) bool {
+			r.UUID = util.UUID()
+			copy(buf[:], r.UUID)
+			_, err = uuidFile.Write(buf)
+			logrus.Infof("wrote new uuid in file with error", err)
```
it will be good if we catch error and return it to the callee ?
Replica's should not continue in case of error. It should be errored out.

kmjayadeep replied:
Will do

utkarshmani1997 commented on replica/replica.go:
```diff
@@ -243,6 +248,32 @@ func IsHeadDisk(diskName string) bool {
+			logrus.Infof("wrote new uuid in file with error", err)
+		} else {
+			uuidFile, _ := os.Open(r.diskPath(uuidFileName))
+			_, err := uuidFile.Read(buf)
```
same comment as earlier.

utkarshmani1997 commented on controller/control.go:
```diff
@@ -317,10 +320,13 @@ func (c *Controller) registerReplica(register types.RegReplica) error {
 func (c *Controller) signalReplica() error {
-	if err := c.factory.SignalToAdd(c.MaxRevReplica, "start"); err != nil {
+	address := c.ReplicaUUIDMap[c.MaxRevReplica]
+	logrus.Infof("Signaling replica for uuid %v and address %v ", c.MaxRevReplica, address)
```
can the logs be changed to "signalling replica: address (uuid) to start"

utkarshmani1997 commented on controller/rest/replica.go:
```diff
@@ -104,11 +104,12 @@ func (s *Server) RegisterReplica(rw http.ResponseWriter, req *http.Request) erro
 		return err
 	}
-	logrus.Infof("Register Replica for address %v", regReplica.Address)
+	logrus.Infof("Register Replica for address %v and uuid %v", regReplica.Address, regReplica.UUID)
```
can the logs be changed to "register replica: address (uuid)"

utkarshmani1997 commented on sync/sync.go:
```diff
@@ -320,9 +322,13 @@ Register:
 		}
 	}
+
+	logrus.Infof("ACTION RECEIVED: %v", action)
```
remove this line

utkarshmani1997 commented on sync/sync.go:
```diff
@@ -320,9 +322,13 @@ Register:
 	if action == "start" {
 		logrus.Infof("Received start from controller")
 		return t.client.Start(replicaAddress)
+	} else if action == "done" {
```
this is not required, after getting 200 OK from controller if replicaCount > 0, replica should continue and perform various tasks such as get added to controller, rebuild, sync etc.

kmjayadeep commented:
I've rebased with master, addressed all review comments and pushed again. Now better code hub is failing, and I'm not able to check what i need to change. Can someone help on this?

utkarshmani1997 commented:
@payes kindly review this PR

kmjayadeep commented:
@utkarshmani1997  Thank you for your support as well :)

utkarshmani1997 commented:
@kmjayadeep one thing i missed completely. You need to do some more work on it. You have to handle Start(addresses...string) ---> Start(uuid.... string) in controller/control.go
Because the main use of UUID starts from the above function.

vishnuitta commented:
are we handling upgrade scenarios?

kmjayadeep commented:
@vishnuitta working on it

kmjayadeep commented:
@payes @utkarshmani1997 I've added UUID to add functions as well (still working on test cases). Can you please give a hit on what all test cases i should add to verify the need of UUIDs?. In the last commit I've already added one test case where maxRevReplica is crashed and restarted with different Ip.

kmjayadeep commented:
Can someone share bettercodehub scan details? I don't have access to it

utkarshmani1997 commented:
Thanks @kmjayadeep for your continuous efforts on this PR. We are going forward with the older design and closing this PR. This PR will be opened if the need of uuid based registration is required.
